horovod/spark/common/util.py

```diff
@@ -0,0 +1,538 @@
+            train_data_path = store.get_train_data_path(dataset_idx)
+            val_data_path = store.get_val_data_path(dataset_idx)
+            if verbose:
+                print('using cached dataframes for key: {}'.format(key))
+                print('train_data_path={}'.format(train_data_path))
+                print('train_rows={}'.format(train_rows))
+                print('val_data_path={}'.format(val_data_path))
+                print('val_rows={}'.format(val_rows))
+        else:
+            train_data_path, val_data_path, dataset_idx = _training_cache.create_data_paths(key, store)
+            if verbose:
+                print('writing dataframes')
+                print('train_data_path={}'.format(train_data_path))
+                print('val_data_path={}'.format(val_data_path))
+
+            schema_cols = feature_columns + label_columns
+            if sample_weight_col:
+                schema_cols.append(sample_weight_col)
+            if isinstance(validation, str):
+                schema_cols.append(validation)
+            df = df[schema_cols]
+
+            metadata = _get_metadata(df)
+
+            to_petastorm = to_petastorm_fn(schema_cols, metadata)
+            train_df = df.rdd.map(to_petastorm).toDF()
+            train_df, val_df, validation_ratio = _train_val_split(train_df, metadata, validation)
+
+            train_partitions = max(int(num_partitions * (1.0 - validation_ratio)),
+                                   num_processes)
+            if verbose:
+                print('train_partitions={}'.format(train_partitions))
+
+            train_df \
+                .coalesce(train_partitions) \
+                .write \
+                .mode('overwrite') \
+                .parquet(train_data_path)
+
+            if val_df:
+                val_partitions = max(int(num_partitions * validation_ratio),
+                                     num_processes)
+                if verbose:
+                    print('val_partitions={}'.format(val_partitions))
+
+                val_df \
+                    .coalesce(val_partitions) \
+                    .write \
+                    .mode('overwrite') \
+                    .parquet(val_data_path)
+
+            train_rows, val_rows, pq_metadata, avg_row_size = get_simple_meta_from_parquet(
+                store, label_columns, feature_columns, sample_weight_col, dataset_idx)
+
+            if verbose:
+                print('train_rows={}'.format(train_rows))
+            if val_df:
+                if val_rows == 0:
+                    raise ValueError(
+                        'Validation DataFrame does not any samples with validation param {}'
+                        .format(validation))
+                if verbose:
+                    print('val_rows={}'.format(val_rows))
+
+            _training_cache.put(key, (train_rows, val_rows, metadata, avg_row_size, dataset_idx))
+
+    try:
+        yield train_rows, val_rows, metadata, avg_row_size, dataset_idx
```

Comment:
The `key` could map to `dataset_idx` and `dataset_idx` to `train_rows, val_rows, metadata`, then `yield dataset_idx` is sufficient here and calling code can get this through extra method `get_dataset_properties(dataset_idx)`

The `key` is deterministic given `df` and `store`. But this determinism is currently not needed, as it maps to a pseudo-random `dataset_idx`. So that `dataset_idx` feels like the actual key that the cache uses. Might all simplify if you consider cross-application `caching` feature.

Reply:
Yes, I like that design better.  I've refactored the cache to have a cleaner separation between keys and dataset indices.